src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,856 @@
+  // sample from thread while in native
+  NATIVE_SAMPLE = 2
+};
+
+static bool thread_state_in_java(JavaThread* thread) {
+  assert(thread != nullptr, "invariant");
+  switch(thread->thread_state()) {
+    case _thread_new:
+    case _thread_uninitialized:
+    case _thread_new_trans:
+    case _thread_in_vm_trans:
+    case _thread_blocked_trans:
+    case _thread_in_native_trans:
+    case _thread_blocked:
+    case _thread_in_vm:
+    case _thread_in_native:
+      break;
+    case _thread_in_Java_trans:
+    case _thread_in_Java:
+      return true;
+    default:
+      ShouldNotReachHere();
+      break;
+  }
+  return false;
+}
+
+static bool thread_state_in_native(JavaThread* thread) {
+  assert(thread != nullptr, "invariant");
+  switch(thread->thread_state()) {
+    case _thread_new:
+    case _thread_uninitialized:
+    case _thread_new_trans:
+    case _thread_in_Java_trans:
+    case _thread_in_Java:
+          break;
+    case _thread_blocked_trans:
+    case _thread_in_vm_trans:
+    case _thread_in_native_trans:
+    case _thread_blocked:
+    case _thread_in_native:
+    case _thread_in_vm: // walking in vm causes weird bugs (assertions in G1 fail), so don't
+      return true;
+    default:
+      ShouldNotReachHere();
+      break;
+  }
+  return false;
+}
+
+static bool is_excluded(JavaThread* thread) {
+  return thread->is_hidden_from_external_view() ||
+    (os::is_readable_pointer(thread->jfr_thread_local()) && thread->jfr_thread_local()->is_excluded());
+}
+
+static JavaThread* get_java_thread_if_valid() {
+  Thread* raw_thread = Thread::current_or_null_safe();
+  JavaThread* jt;
+
+  if (raw_thread == nullptr || !raw_thread->is_Java_thread() ||
+      (jt = JavaThread::cast(raw_thread))->is_exiting()) {
+    return nullptr;
+  }
+
+  if (is_excluded(jt)) {
+    return nullptr;
+  }
+  return jt;
+}
+
+// A trace of stack frames, contains all information
+// collected in the signal handler, required to create
+// a JFR event with a stack trace
+class JfrCPUTimeTrace {
+  friend class JfrTraceQueue;
+  u4 _index;
+  JfrAsyncStackFrame* _frames;
+  JfrAsyncStackTrace _stacktrace;
+  u4 _max_frames;
+  // error code for the trace, 0 if no error
+  u4 _error;
+
+  JfrSampleType _type;
+  JfrTicks _start_time;
+  JfrTicks _end_time;
+  JavaThread* _sampled_thread;
+
+public:
+  JfrCPUTimeTrace(u4 index, JfrAsyncStackFrame* frames, u4 max_frames):
+    _index(index), _frames(frames), _stacktrace(_frames, max_frames),
+    _max_frames(max_frames) {
+
+    }
+
+  JfrAsyncStackFrame* frames() { return _frames; }
+  u4 max_frames() const { return _max_frames; }
+
+  bool successful() const { return _error == NO_ERROR; }
+
+  JfrSampleType type() const { return _type; }
+
+  JfrTicks start_time() const { return _start_time; }
+  void set_end_time(JfrTicks end_time) { _end_time = end_time; }
+  JfrTicks end_time() const { return _end_time; }
+  void set_sampled_thread(JavaThread* thread) { Atomic::store(&_sampled_thread, thread); }
```

Comment:
Why atomic store? Again, try to use the traceid of the thread (from the JfrThreadLocal (updated dynamically depending on vthread or regular thread).